OncoKB icon tooltip support

onursumer opened the pull request:
# What? Why?

Added tool tip (OncoKB Card) to the annotation column for the OncoKB icon.

Note that currently `OncoKbUtils` class contains a lot of legacy code. I just made it work with TypeScript with minimal changes. We need to refactor both `OncoKBCard` and `OncoKBUtils` in a more modular and testable way.

# Checks
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

adufilie commented on src/pages/patientView/PatientViewPage.tsx:
```diff
@@ -482,6 +482,7 @@ export default class PatientViewPage extends React.Component<IPatientViewPagePro
                                     cosmicData={this.state.cosmicData}
                                     mrnaExprRankData={ patientViewPageStore.mrnaExprRankCache.cache }
                                     oncoKbData={patientViewPageStore.oncoKbData.isComplete ? patientViewPageStore.oncoKbData.result : undefined}
+                                    pmidData={patientViewPageStore.pmidData.isComplete ? patientViewPageStore.pmidData.result : undefined}
```
These ternary operators are unnecessary

onursumer replied:
Good to know!

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
                 // but the actual return type is {meta: {} data: Array<IndicatorQueryResp>}
                 // that's why here we need to force data type to be any and get actual data by data.data
-                oncokbSearchPromise.then((data:any) => {
+                Promise.all([oncokbSearchPromise, evidenceLookupPromise]).then((values) => {
```
Use destructuring here:
`let [onkokbSearch, evidenceLookup] = await Promise.all([oncokbSearchPromise, evidenceLookupPromise]);`

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
                     sampleIdToTumorType[mutation.sampleId]);
             }), "id");
 
-            return new Promise((resolve, reject) => {
```
This code would be much cleaner with async/await

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
                     const oncoKbData:IOncoKbData = {
                         sampleToTumorMap: sampleIdToTumorType,
-                        indicatorMap: generateIdToIndicatorMap(data.data)
+                        indicatorMap: generateIdToIndicatorMap((values[0] as any).data),
```
Instead of letting the `any` propagate throughout the whole function, either fix the generated API code to have the correct type or cast the promise as the correct type. You can do `as any as CorrectType`.

onursumer replied:
Long term plan is to fix this problem in the actual OncoKB API definition. Return types are not defined correctly there.

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
+        }
+    }, {sampleToTumorMap: {}, indicatorMap: {}, evidenceMap: {}});
+
+    readonly pmidData = remoteData<any>({
```
It's better to let typescript determine the type instead of using `any`

adufilie replied:
I see, this would be an implicit any.  So we should define the correct type instead.

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
+            const evidenceLookupPromise = oncokbClient.evidencesLookupPostUsingPOST(
+                {body: generateEvidenceQuery(queryVariants)});
 
+            return new Promise((resolve, reject) => {
```
This code would be cleaner with async/await

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
+        invoke: () => {
+            let refs:any[] = [];
+
+            _.each(this.oncoKbData.result.evidenceMap, (evidence:any) => {
```
Make the result have the correct type instead of using `any` everywhere.

adufilie commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -341,27 +341,104 @@ export class PatientViewPageStore
+                }
+            });
+
+            return new Promise((resolve, reject) => {
```
This would be cleaner with async/await.

adufilie commented on src/pages/patientView/mutation/column/AnnotationColumnFormatter.tsx:
```diff
@@ -37,6 +38,8 @@ export interface IAnnotation {
     is3dHotspot: boolean;
     myCancerGenomeLinks: string[];
     oncoKbIndicator?: IndicatorQueryResp;
+    oncoKbEvidence?: any;
```
Try to avoid `any`

onursumer commented:
@adufilie I agree that we need proper types. I don't like seeing `any`s all over the place either, but the idea for this sprint was to make the legacy OncoKB code work with the current framework.

@alisman do you think we should try addressing typing issues in this PR, or is it okay to deal with them in the future?

onursumer commented:
@adufilie @alisman: updated the PR with Andy's suggestions. Also added a partial model for `Evidence` (just copied from the model from legacy code), and got rid of some of the `any`s. Now the code looks slightly better.

onursumer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(13) Mutation Table --> oncokb mouseover card](https://trello.com/c/adbegZpf/36-13-mutation-table-oncokb-mouseover-card)

onursumer commented:
@adamabeshouse yes, we should definitely defer fetching of the tooltip data until mouseover event. But it should be okay to do it in another PR. We should also move related data fetching functions from `PatientViewPage` to `PatientViewStore`.
